### PR TITLE
[Runtime] Add hooks for back-deploying dynamic casting and type lookup changes

### PR DESCRIPTION
**Pull request by mikeash**
We anticipate expanding the capabilities of dynamic casting in future releases. For a big example, we might add support for dynamic casting to CF types. With the goal of ABI stability drawing nearer, we want to allow apps compiled with future compilers to still take advantage of these changes when running on older runtimes. This PR accomplishes that by providing hooks at the appropriate runtime entrypoints that apps can use to substitute in their own functionality.

The pertinent design points are:

* Hook points are added to the various `swift_dynamicCast` functions, plus `swift_conformsToProtocol` and `swift_getTypeByMangledName`.
* On the first call of one of these functions, the hooking mechanism searches for a section called `__swift_lite` in the executable's `__DATA` segment. This section contains function pointers for the various hooks.
* If the `__swift_lite` section exists *and* contains a non-`NULL` hook for the function being called, the function calls the hook. The hook is passed a function pointer to the original implementation, so that it may invoke the original implementation at the time of its choosing.
* If the hook is `NULL` or the `__swift_lite` section does not exist, the original implementation is called directly.
* To avoid confusion and conflict, *only* the executable is searched. Libraries are not searched. If any library has a `__swift_lite` section, that section is ignored.

I hope I've added hooks to all the entrypoints that might need them, but please double-check to see if I missed anything important.

rdar://problem/33843263

### CODE REVIEW

**Comment by jckarter** [comment]
I wonder whether some of the repetition here could be avoided using an xmacro file along the lines of `include/swift/Runtime/RuntimeFunctions.def`. It seems like you could use that to splat out the data structure, the dispatch thunks, and the typedefs from one data source, and would also make it easier to add more overrides here in the future.

Since the override implementations themselves should end up linked in the binary with the dispatch table, I think we could also use `RelativeDirectPointer`s to the override implementations in the `__swift_lite` table, which would make it smaller, and let us put it in `__TEXT` so that it's in constant memory.

File: stdlib/public/runtime/ImageInspectionCOFF.cpp [hunk 1]
```diff
@@ -139,4 +139,10 @@ int swift::lookupSymbol(const void *address, SymbolInfo *info) {
 #endif // __CYGWIN__
 }
 
+// This is only used for backward deployment hooks, which we currently only support for
+// MachO. Add a stub here to make sure it still compiles.
+void *lookupSection(const char *segment, const char *section, size_t *outSize) {
+  return NULL;
```
Comment (gparker42) [inline stdlib/public/runtime/ImageInspectionCOFF.cpp]: s/NULL/nullptr/

**Comment by mikeash** [comment]
@jckarter I tried not to go overboard with the macros so as not to obscure what's going on too much. Maybe I'll try it anyway and see how it looks.

I was thinking that it would be nice to allow the hooks to be in a dylib for targets where that's suitable. Obviously this wouldn't be required, since we want to support standalone executables, but it could be a good option to have.

**Comment by mikeash** [comment]
@jckarter Take a look at the stuff I just pushed. I think your suggestion ended up looking pretty nice.

**Comment by jckarter** [comment]
Yeah, looks pretty good!

**Comment by jrose-apple** [comment]
I know I'm late commenting on this, but should the `__swift_lite` section have a version field? That way we could in theory pick the highest version, instead of relying on the executable.

**Comment by mikeash** [comment]
@jrose-apple There actually is a version field, although the intent is different. I don't have a concrete use case for it just yet, but the idea is that a distant future Swift runtime will be able to look at the version field to determine whether the section was targeted at it or at an older runtime.

We're deliberately only looking in the executable just to keep things simple. We can come up with schemes to pick hooks from libraries as well, but that can get complicated fast, and by the time we realize this scheme is broken it'll probably be too late to fix it. If we ever use these hooks, library code will have to be a bit more restricted not to take advantage of them, or will have to require being bundled into an app that has them.

Good point on the name. How about `__swift_hooks`? Or if you have another idea I'm open to suggestions.

**Comment by jrose-apple** [comment]
I conflated two things with the "version" idea. The more important part is what you said: if the system Swift is newer than the executable Swift, maybe it doesn't want to use the hooks. Of course, that rules out ad hoc hooks that are always supposed to win.

`__swift_hooks` sounds fine to me; the one I was toying with last night was `__swift_runtime_overrides` but that's kind of long.

**Comment by mikeash** [comment]
I believe Mach-O has a limit of 16 bytes for section names, too.

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
@swift-ci please test

**Pull request merged**